### PR TITLE
Remove dst file on failed network download

### PR DESCRIPTION
**Pull request by dmcilvaney**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [ ] The toolchain has been rebuilt successfully (or no changes were made to it)
- [ ] The toolchain/worker package manifests are up-to-date
- [ ] Any updated packages successfully build (or no packages were changed)
- [ ] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [ ] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [ ] All package sources are available
- [ ] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [ ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [ ] All source files have up-to-date hashes in the `*.signatures.json` files
- [ ] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [ ] Documentation has been updated to match any changes to the build system
- [ ] Ready to merge

---

###### Summary 

When `network.DownloadFile()` it was leaving files on the disk. The pre-cacher tool would create empty `.rpm` files which would confuse the fetcher and scheduler further downstream.

###### Change Log  

- network.DownloadFile() will now clean up its dst file on error
- Added `file.RemoveFileIfExists()`
- Add file_test.go 

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

- Local test

### CODE REVIEW

File: toolkit/tools/graphpkgfetcher/graphpkgfetcher.go [hunk 1]
```diff
@@ -423,6 +423,8 @@ func downloadSingleDeltaRPM(realDependencyGraph *pkggraph.PkgGraph, buildNode *p
 			logger.Log.Warnf("Can't find delta RPM to download for %s: %s (local copy may be newer than published version)", fullyQualifiedRpmName, err)
 			return nil
 		}
+	} else {
```
Comment (neha170) [inline toolkit/tools/graphpkgfetcher/graphpkgfetcher.go:426]: This seems unrelated to the PR, and probably deserves its own PR. I am thinking from the perspective where we might want to roll-back in some imaginary scenario, and would end up erroneously deleting this edge case in graphpkgfetcher.

File: toolkit/tools/internal/network/network.go [hunk 2]
```diff
@@ -65,6 +75,22 @@ func DownloadFile(url, dst string, caCerts *x509.CertPool, tlsCerts []tls.Certif
...
+	if exists {
+		err = os.Remove(fullPath)
+		if err != nil {
+			err = fmt.Errorf("failed to remove file:\n%w", err)
+			return
+		}
+	}
+	return
+}
+
```
Comment (PawelWMS) [inline toolkit/tools/internal/network/network.go:93]: I think we might not even need to check if the file exists and rely on `os.Remove` error entirely by ignoring a missing file:
```suggestion
func removeFileIgnoreMissing(fullPath string) (err error) {
	removeErr := os.Remove(fullPath).(*os.PathError)
	if removeErr != nil && removeErr.Err != syscall.ENOENT {
		return fmt.Errorf("failed to remove file (%s):\n%w", fullPath, removeErr)
	}
	return removeErr
}
```

And what do you think about making that a function inside `file.go`?

File: toolkit/tools/internal/network/network.go [hunk 3]
```diff
@@ -30,14 +31,23 @@ func JoinURL(baseURL string, extraPaths ...string) string {
...
+	defer func() {
+		// If there was an error, ensure that the file is removed
+		if err != nil {
+			cleanupErr := removeFileIfExists(dst)
+			if cleanupErr != nil {
+				logger.Log.Errorf("Failed to remove failed network download file '%s': %s", dst, err)
+			}
+		}
+	}()
 	defer dstFile.Close()
```
Comment (PawelWMS) [inline toolkit/tools/internal/network/network.go:51]: Now the file will get closed before it gets removed, so there's a window where another thread/process may create a new file, which gets then removed by this code. I'd reverse the order of operations. Also, I think there's no need for two separate `defer` if they are called immediately one after another - we can lump these together.

File: toolkit/tools/graphpkgfetcher/graphpkgfetcher.go [hunk 4]
```diff
@@ -423,6 +423,8 @@ func downloadSingleDeltaRPM(realDependencyGraph *pkggraph.PkgGraph, buildNode *p
...
+		logger.Log.Debugf("Found pre-cached delta RPM for %s, skipping download", fullyQualifiedRpmName)
 	}
 
 	foundCacheRPM, err = file.PathExists(cachedRPMPath)
```
Comment (PawelWMS) [inline toolkit/tools/graphpkgfetcher/graphpkgfetcher.go:430]: If we're introducing the `else`, then we should move the `foundCacheRPM, err = file.PathExists(cachedRPMPath)` call and repeated `foundCacheRPM` into the `if !foundCacheRPM {` block above. Then, if the cached RPM is still not there after we try to clone it, we could exit early with the `logger.Log.Warnf("Delta download for '%s' did not generate the correct delta RPM: '%s'", buildNode, cachedRPMPath)` warning we have below.

File: toolkit/tools/graphpkgfetcher/graphpkgfetcher.go (hunk 1 above)
Comment (dmcilvaney) [reply toolkit/tools/graphpkgfetcher/graphpkgfetcher.go:426]: This doesn't materially affect the behavior of the tool, so I don't mind bundling it (and it was the debugging information that allowed me to diagnose the issue this is solving so I feel its at least partially related).

File: toolkit/tools/internal/network/network.go (hunk 2 above)
Comment (dmcilvaney) [reply toolkit/tools/internal/network/network.go:93]: Moved to file.go, and as per comments in the go library `if removeErr != nil && !errors.Is(removeErr, fs.ErrNotExist) {` is the modern way to check for this error.

File: toolkit/tools/graphpkgfetcher/graphpkgfetcher.go (hunk 4 above)
Comment (dmcilvaney) [reply toolkit/tools/graphpkgfetcher/graphpkgfetcher.go:430]: I'll make a todo, but I'd like to limit extra changes here.

File: toolkit/tools/internal/file/file_test.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	logger.InitStderrLog()
+	retVal := m.Run()
+	os.Exit(retVal)
+}
+
+func testFileName(t *testing.T) string {
+	return filepath.Join(t.TempDir(), t.Name())
+}
+
+func TestRemoveFileIfExistsValid(t *testing.T) {
```
Comment (PawelWMS) [inline toolkit/tools/internal/file/file_test.go:25]: Thanks for adding the tests! Could we also add a check that the removed file does not exist after calling `RemoveFileIfExists`?

**Review by PawelWMS — APPROVED** [review]

File: toolkit/tools/internal/file/file_test.go (hunk 5 above)
Comment (dmcilvaney) [reply toolkit/tools/internal/file/file_test.go:27]: Realized I wasn't using t.TempDir right, fixed that as well.
